packages/server/src/app-server.ts

```diff
@@ -1,39 +1,83 @@
-import { timeout, Operation } from 'effection';
-import { Mailbox } from '@bigtest/effection';
+import { timeout, Operation, Context, fork } from 'effection';
+import { Subscribable } from '@effection/subscription'
 import { fetch } from '@effection/fetch';
 import { ChildProcess } from '@effection/node';
 import * as process from 'process';
+import { OrchestratorState, AppServiceState, AppOptions, AppStatus } from './orchestrator/state';
+import { Slice } from '@bigtest/atom';
 
 interface AppServerOptions {
-  delegate: Mailbox;
-  dir?: string;
-  command: string;
-  args?: string[];
+  slice: Slice<AppServiceState, OrchestratorState>;
+  url: string;
+  command?: string;
   env?: Record<string, string>;
-  port: number;
+  dir?: string;
 };
 
-function* isReachable(port: number) {
+function* isReachable(url: string) {
   try {
-    let response: Response = yield fetch(`http://127.0.0.1:${port}`)
-    return response.ok
+    let response: Response = yield fetch(url);
+    return response.ok;
   } catch (error) {
     return false;
   }
 }
 
-export function* createAppServer(options: AppServerOptions): Operation {
-  yield ChildProcess.spawn(options.command, options.args || [], {
-    cwd: options.dir,
-    detached: true,
-    env: Object.assign({}, process.env, options.env),
-  });
+export function* createAppServer({ slice, ...options }: AppServerOptions): Operation {
+  let appOptions = slice.slice<AppOptions>(['appOptions']);
+  let appStatus = slice.slice<AppStatus>(['appStatus']);
+  let current: Context;
+
+  function* startApp(appOptions: AppOptions): Operation<void> {
+    if (!appOptions) {
+      return;
+    }
+
+    if (current) {
+      console.debug("[app]", 'halting');
+      current.halt();
+    }
+
+    console.debug("[app]", 'unstarted');
+    appStatus.set('unstarted');
+
+    current = yield fork(function* () {
+      if (appOptions.command) {
+        let child = yield ChildProcess.spawn(appOptions.command as string, [], {
+          cwd: appOptions.dir,
+          detached: true,
+          env: Object.assign({}, process.env, appOptions.env),
+          stdio: 'inherit'
+        });
+
+        child.on('exit', (code: number, signal: string) => {
```

Comment:
We don't want to use a direct event listener here, because it will never get uninstalled. Probably not much of a difference in this case since the listere and the child process have the same scope of life. In general though, for all `EventEmitter` and `EventTarget` objects, you can use the `on` method from `@effection/events`. e.g.

```ts
let [code, signal]: [number, string] = yield on(child, 'exit');
console.log("🧨", code, signal);
appStatus.set('crashed');
```

Not only do you have your lexical context freed from a closure, but you never have to worry about having to call `off()` because effection will do it for you.